lib/Server.js

```diff
@@ -81,861 +83,928 @@ const STATS = {
+      );
+      res.write(_path);
+      res.write('.js');
+      res.write(req._parsedUrl.search || '');
+
+      res.end('"></script></body></html>');
+    } catch (err) {
+      return next();
+    }
   }
-};
 
-Server.prototype._watch = function(watchPath) {
-  // duplicate the same massaging of options that watchpack performs
-  // https://github.com/webpack/watchpack/blob/master/lib/DirectoryWatcher.js#L49
-  // this isn't an elegant solution, but we'll improve it in the future
-  const usePolling = this.watchOptions.poll ? true : undefined;
-  const interval =
-    typeof this.watchOptions.poll === 'number'
-      ? this.watchOptions.poll
-      : undefined;
-
-  const options = {
-    ignoreInitial: true,
-    persistent: true,
-    followSymlinks: false,
-    depth: 0,
-    atomic: false,
-    alwaysStat: true,
-    ignorePermissionErrors: true,
-    ignored: this.watchOptions.ignored,
-    usePolling,
-    interval,
-  };
+  // send stats to a socket or multiple sockets
+  _sendStats(sockets, stats, force) {
+    if (
+      !force &&
+      stats &&
+      (!stats.errors || stats.errors.length === 0) &&
+      stats.assets &&
+      stats.assets.every((asset) => !asset.emitted)
+    ) {
+      return this.sockWrite(sockets, 'still-ok');
+    }
 
-  const watcher = chokidar.watch(watchPath, options);
+    this.sockWrite(sockets, 'hash', stats.hash);
 
-  watcher.on('change', () => {
-    this.sockWrite(this.sockets, 'content-changed');
-  });
+    if (stats.errors.length > 0) {
+      this.sockWrite(sockets, 'errors', stats.errors);
+    } else if (stats.warnings.length > 0) {
+      this.sockWrite(sockets, 'warnings', stats.warnings);
+    } else {
+      this.sockWrite(sockets, 'ok');
+    }
+  }
 
-  this.contentBaseWatchers.push(watcher);
-};
+  _watch(watchPath) {
+    // duplicate the same massaging of options that watchpack performs
+    // https://github.com/webpack/watchpack/blob/master/lib/DirectoryWatcher.js#L49
+    // this isn't an elegant solution, but we'll improve it in the future
+    const usePolling = this.watchOptions.poll ? true : undefined;
+    const interval =
+      typeof this.watchOptions.poll === 'number'
+        ? this.watchOptions.poll
+        : undefined;
+
+    const options = {
+      ignoreInitial: true,
+      persistent: true,
+      followSymlinks: false,
+      depth: 0,
+      atomic: false,
+      alwaysStat: true,
+      ignorePermissionErrors: true,
+      ignored: this.watchOptions.ignored,
+      usePolling,
+      interval,
+    };
+
+    const watcher = chokidar.watch(watchPath, options);
+
+    watcher.on('change', () => {
+      this.sockWrite(this.sockets, 'content-changed');
+    });
 
-Server.prototype.invalidate = function() {
-  if (this.middleware) {
-    this.middleware.invalidate();
+    this.contentBaseWatchers.push(watcher);
   }
-};
+
+  invalidate() {
+    if (this.middleware) {
+      this.middleware.invalidate();
+    }
+  }
+}
```

Comment:
Let's merge https://github.com/webpack/webpack-dev-server/pull/1658 first and when rebase :+1: